src/common/config.c

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright (C) 2020 Evan Lojewski
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#include <board/flash.h>
+#include <common/config.h>
+#include <common/debug.h>
+#include <stddef.h>
+#include <string.h>
+
+#define CONFIG_FLASH_SIZE       (4 * 1024)      /* 4K sector to allow easy erasing */
+
+// Prevent failures to compile on AVR
+#ifndef __SDCC
+    #define __code
+#endif
+
+config_t *gConfigEntries = NULL;
+
+#define CONFIG_ADDRESS  (0x10000L)
+#define CONFIG_SIZE   (1 * 1024)  /* 1K block - code currently assumes this matches the erase size */
+#define CONFIG_PERSISTANT_MAGIC     (0xCF1C0F10)
+
+uint32_t __code __at(CONFIG_ADDRESS) config_magic;
+uint8_t __code __at(CONFIG_ADDRESS + sizeof(config_magic)) config_persistant[CONFIG_SIZE - sizeof(config_magic)];
+
+#define CONFIG_FLASH_ENTRY_LENGTH_INVALID   (0xff)
+#define CONFIG_FLASH_ENTRY_CURRENT          (0xFF)
+#define CONFIG_FLASH_ENTRY_OBSOLETE         (0x00)
+
+#define CONFIG_FLASH_ENTRY_CURRENT_OFFSET   (0)
+#define CONFIG_FLASH_ENTRY_LEN_OFFSET       (1)
+#define CONFIG_FLASH_ENTRY_VALUE_OFFSET     (2)
+#define CONFIG_FLASH_ENTRY_NAME_OFFSET      (6)
+#define CONFIG_FLASH_ENTRY_HEADER_SIZE      (6)
+// config_persistant format
+// (1) byte: Current / Obsolete. When a config entry is updated, byte is written to Obsolete and a new entry is added later.
+// (1) byte: Short Name Length. When 0xFF, no more config entries exist.
+// (4) bytes: Value
+// Len Bytes: short name.
+
+int8_t flash_strncmp(uint32_t addr, const char* str, uint8_t len) {
+    while(len) {
+        char byte = flash_read_u8(addr);
+
+        char res = *str - byte;
+        if(res) {
+            return res;
+        }
+        len--;
+        addr++;
+        str++;
+    }
+
+    return 0;
+}
+
+
+bool config_validate_magic(void) {
+    uint32_t magic = flash_read_u32(CONFIG_ADDRESS);
+
+    return (CONFIG_PERSISTANT_MAGIC == magic);
+}
+
+uint32_t config_find_free_entry() {
+    if (!config_validate_magic()) {
+        // Write magic.
+        flash_write_u32(CONFIG_ADDRESS, CONFIG_PERSISTANT_MAGIC);
+    }
+
+    uint32_t addr = CONFIG_ADDRESS + sizeof(config_magic);
+    uint32_t limit = addr + sizeof(config_persistant);
+    while(addr < limit) {
+        // configuration present.
+        uint8_t len;
+        len = flash_read_u8(addr + CONFIG_FLASH_ENTRY_LEN_OFFSET);
+
+        if(CONFIG_FLASH_ENTRY_LENGTH_INVALID == len) {
+            // Free entry found.
+            return addr;
+        }
+
+        // Go to the next entry.
+        addr = addr + CONFIG_FLASH_ENTRY_NAME_OFFSET + len;
```

Comment:
With SDCC, this actually results in two different binaries being produced. From my initial testing, "addr = addr +" appeared to alleviate some memory issues I was running into (since improves with the i2c change).. 

Form my retesting today, there's actually a 5 byte hit in code size when using addr = addr + instead of +=.
I've updated this.